File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number} target
 * @return {number}
 */

let getTarget = (targetSum, numOfDice, k, memo = {}) => {

    // dp based on numOfDice + targetSum
    let key = targetSum + "-" + numOfDice

    if(key in memo)
      return memo[key]

    if(targetSum === 0 &&  numOfDice == 0)
        return 1
    if(targetSum < 0 || numOfDice <= 0)
        return 0
    
    let ans = 0
    for(let i = 1; i <= k; i++) {
        ans = (ans + getTarget(targetSum - i, numOfDice - 1, k, memo)) % (10**9 + 7);
    }

    memo[key] = ans
    return ans
}

var numRollsToTarget = function(n, k, target) {
  return getTarget(target, n, k)
};

File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.java
class Solution {
    Map<String, Integer> memo = new HashMap<>();
    
    public int numRollsToTarget(int n, int k, int target) {
        int ans = 0;
        
        String key = n + "_" + target;
        
        if(memo.containsKey(key))
            return memo.get(key);
        
        // desired state, this is one of the possible ways
        if(n == 0 && target == 0)
            return 1;
        
        // dice are over but target is not met
        // or we still have dice and target becomes invalid
        // not a valid way, reject it
        if(target < 0 || n <= 0)
            return 0;
        
        for(int i=0; i<k; i++) {
            ans = (ans + numRollsToTarget(n-1, k, target-(i+1))) % 1000000007; 
        }
        
        memo.put(key, ans);
        
        return ans;
    }
}